### ISSUE TITLE
Replace recursion with fixpoints and case trees

### ISSUE THREAD

**Issue by AndrasKovacs** [AndrasKovacs/cctt#15]
The current recursion setup has a problem: a recursive name is not convertible to the result of normalizing it. Evaluation unfolds the symbol once, then quotation does not unfold recursive calls anymore. So overall the "normal form" of a recursive name is a one-step unfolding. Currently, it is possible to produce code by quotation which is not checkable.

```
add (x y : ℕ) : ℕ := case x [
  zero . y;
  suc x. suc (add x y)];

test : add = add := refl;
```
The normal form of `test` does not check at its original type because it unfolds `add`. 

This problem already exists in `mini-tt`  which inspired the recursion implementation here. [PiSigma](https://www.cs.nott.ac.uk/~psztxa/publ/pisigma-new.pdf) doubles down on recursive definitions but it adds much extra complexity, and I'm not actually sure that it has a complete solution to the above issue.

Proposal: we add a primitive N-ary fixpoint. Every fixpoint definition formally starts with a *case tree*, more concretely a nested lambda-case. A fixpoint application computes when it is fully applied and we can compute past the case tree. *In addition*, we keep
around the primitive non-recursive case splits. 

- Coq fixpoints are not formally associated to a case tree. This might seem more flexible, but the rules for computing a fixpoint application become much uglier. A Coq fixpoint only computes if it's applied to a "canonical" value for its structurally decreasing argument. Hence, we need to pick a structural argument before we can compute. Also, lexicographic recursion does not work out of the box. 
- Agda has no fixpoints, only top-level recursive definitions which are associated to case trees. Computing applications work the same as in my proposed setup, but we can't write nested or local recursive definitions, can't compare bodies of case-splitting definitions for conversion, and, out of the box, can't produce normalized bodies of recursive functions in ordinary quotation. 

In short, the proposed fixpoints integrate Coq-style, in that fixpoints are first-class, and Agda-style, in that structural recursion must be defined with a formal case tree.
